File: react_client/src/blocs/map/map_bloc.ts
// ABOUTME: BLoC for managing map visualization state.
// ABOUTME: Handles adding/removing geographic features and auto-framing.

import { Bloc } from "@granite-crisp/react-bloc";
import type { MapEvent } from "./map_event";
import {
  type MapState,
  type MapMarker,
  type MapPolygon,
  type LatLng,
  type LatLngBounds,
  createInitialMapState,
} from "./map_state";
import type { GeoFeature } from "../../models";

// Auto-frame thresholds
const SINGLE_POINT_THRESHOLD = 0.0001; // ~11 meters at equator
const SINGLE_POINT_PADDING = 0.02; // ~2km view for single points
const NORMAL_BOUNDS_PADDING = 0.25; // 25% padding for feature collections
const VIEW_ADJUSTMENT_THRESHOLD = 0.3; // Adjust if new features extend >30% beyond view
const MIN_RANGE_FOR_CALCULATION = 0.000001; // Prevent division by zero

export class MapBloc extends Bloc<MapEvent, MapState> {
  constructor() {
    super(createInitialMapState());
  }

  add(event: MapEvent): void {
    switch (event.type) {
      case "addGeoFeature":
        this.handleAddGeoFeature(event.feature);
        break;
      case "removeGeoFeature":
        this.handleRemoveGeoFeature(event.featureId);
        break;
      case "clearMap":
        this.emitState(createInitialMapState());
        break;
      case "zoomIn":
        this.emit({
          zoom: this.state.zoom + 1,
          autoFrameEnabled: false,
        });
        break;
      case "zoomOut":
        this.emit({
          zoom: this.state.zoom - 1,
          autoFrameEnabled: false,
        });
        break;
      case "setZoom":
        this.emit({ zoom: event.zoom });
        break;
      case "updateMapPosition":
        this.emit({
          center: event.center,
          zoom: event.zoom,
        });
        break;
      case "disableAutoFrame":
        this.emit({ autoFrameEnabled: false });
        break;
      case "enableAutoFrame":
        this.handleEnableAutoFrame();
        break;
    }
  }

  private handleAddGeoFeature(feature: GeoFeature): void {
    if (feature.type === "polygon" && feature.coordinates) {
      this.handleAddPolygon(feature);
    } else if (feature.lat !== undefined && feature.lon !== undefined) {
      this.handleAddMarker(feature);
    }
  }

  private handleAddMarker(feature: GeoFeature): void {
    const marker: MapMarker = {
      id: feature.id,
      position: { lat: feature.lat!, lng: feature.lon! },
      label: feature.label,
      category: feature.category,
    };

    const updatedMarkers = [...this.state.markers, marker];
    const allPoints = [
      ...updatedMarkers.map((m) => m.position),
      ...this.extractAllPolygonPoints(),
    ];
    const newPoints = [marker.position];

    this.handleAutoFrame(updatedMarkers, this.state.polygons, allPoints, newPoints);
  }

  private handleAddPolygon(feature: GeoFeature): void {
    const polygon: MapPolygon = {
      id: feature.id,
      coordinates: feature.coordinates!,
      label: feature.label,
      category: feature.category,
    };

    const updatedPolygons = [...this.state.polygons, polygon];
    const newPoints = this.extractPolygonPoints(feature.coordinates!);
    const allPoints = [
      ...this.state.markers.map((m) => m.position),
      ...this.extractAllPolygonPoints(),
      ...newPoints,
    ];

    this.handleAutoFrame(this.state.markers, updatedPolygons, allPoints, newPoints);
  }

  private extractPolygonPoints(coordinates: [number, number][][]): LatLng[] {
    // Extract points from outer ring for bounds calculation
    // GeoJSON coordinates are [lon, lat], we need {lat, lng}
    return coordinates[0].map(([lng, lat]) => ({ lat, lng }));
  }

  private extractAllPolygonPoints(): LatLng[] {
    return this.state.polygons.flatMap((p) => this.extractPolygonPoints(p.coordinates));
  }

  private handleRemoveGeoFeature(featureId: string): void {
    const updatedMarkers = this.state.markers.filter((m) => m.id !== featureId);
    const updatedPolygons = this.state.polygons.filter((p) => p.id !== featureId);
    this.emit({ markers: updatedMarkers, polygons: updatedPolygons });
  }

  private handleAutoFrame(
    updatedMarkers: MapMarker[],
    updatedPolygons: MapPolygon[],
    allPoints: LatLng[],
    newPoints: LatLng[]
  ): void {
    if (!this.state.autoFrameEnabled || newPoints.length === 0) {
      this.emit({ markers: updatedMarkers, polygons: updatedPolygons });
      return;
    }

    const allBounds = this.calculateBoundsFromPoints(allPoints);
    const newBounds = this.calculateBoundsFromPoints(newPoints);

    if (!allBounds || !newBounds) {
      this.emit({ markers: updatedMarkers, polygons: updatedPolygons });
      return;
    }

    // Adjust view if: (1) first features in conversation, or (2) new features significantly outside current view
    const shouldFit =
      this.state.conversationBounds === null ||
      this.shouldAdjustView(newBounds);

    if (shouldFit) {
      const paddedBounds = this.padBounds(allBounds);
      const center = {
        lat: (paddedBounds.southWest.lat + paddedBounds.northEast.lat) / 2,
        lng: (paddedBounds.southWest.lng + paddedBounds.northEast.lng) / 2,
      };

      this.emit({
        markers: updatedMarkers,
        polygons: updatedPolygons,
        conversationBounds: allBounds,
        center,
        fitBoundsVersion: this.state.fitBoundsVersion + 1,
      });
    } else {
      this.emit({
        markers: updatedMarkers,
        polygons: updatedPolygons,
        conversationBounds: allBounds,
      });
    }
  }

  private handleEnableAutoFrame(): void {
    const allPoints = [
      ...this.state.markers.map((m) => m.position),
      ...this.extractAllPolygonPoints(),
    ];
    const conversationBounds = this.calculateBoundsFromPoints(allPoints);

    if (conversationBounds) {
      const paddedBounds = this.padBounds(conversationBounds);
      const center = {
        lat: (paddedBounds.southWest.lat + paddedBounds.northEast.lat) / 2,
        lng: (paddedBounds.southWest.lng + paddedBounds.northEast.lng) / 2,
      };

      this.emit({
        autoFrameEnabled: true,
        conversationBounds,
        center,
        fitBoundsVersion: this.state.fitBoundsVersion + 1,
      });
    } else {
      this.emit({ autoFrameEnabled: true });
    }
  }

  private calculateBoundsFromPoints(points: LatLng[]): LatLngBounds | null {
    if (points.length === 0) return null;

    let minLat = points[0].lat;
    let maxLat = points[0].lat;
    let minLng = points[0].lng;
    let maxLng = points[0].lng;

    for (const point of points) {
      minLat = Math.min(minLat, point.lat);
      maxLat = Math.max(maxLat, point.lat);
      minLng = Math.min(minLng, point.lng);
      maxLng = Math.max(maxLng, point.lng);
    }

    return {
      southWest: { lat: minLat, lng: minLng },
      northEast: { lat: maxLat, lng: maxLng },
    };
  }

  private shouldAdjustView(newFeatureBounds: LatLngBounds): boolean {
    if (!this.state.conversationBounds) return true;

    const currentBounds = this.state.conversationBounds;
    const latRange = currentBounds.northEast.lat - currentBounds.southWest.lat;
    const lngRange = currentBounds.northEast.lng - currentBounds.southWest.lng;

    // If current view is essentially a point, always adjust
    if (
      Math.abs(latRange) < MIN_RANGE_FOR_CALCULATION ||
      Math.abs(lngRange) < MIN_RANGE_FOR_CALCULATION
    ) {
      return true;
    }

    const latOutside =
      (newFeatureBounds.southWest.lat < currentBounds.southWest.lat
        ? currentBounds.southWest.lat - newFeatureBounds.southWest.lat
        : 0) +
      (newFeatureBounds.northEast.lat > currentBounds.northEast.lat
        ? newFeatureBounds.northEast.lat - currentBounds.northEast.lat
        : 0);

    const lngOutside =
      (newFeatureBounds.southWest.lng < currentBounds.southWest.lng
        ? currentBounds.southWest.lng - newFeatureBounds.southWest.lng
        : 0) +
      (newFeatureBounds.northEast.lng > currentBounds.northEast.lng
        ? newFeatureBounds.northEast.lng - currentBounds.northEast.lng
        : 0);

    return (
      latOutside / latRange > VIEW_ADJUSTMENT_THRESHOLD ||
      lngOutside / lngRange > VIEW_ADJUSTMENT_THRESHOLD
    );
  }

  private padBounds(bounds: LatLngBounds): LatLngBounds {
    const latRange = bounds.northEast.lat - bounds.southWest.lat;
    const lngRange = bounds.northEast.lng - bounds.southWest.lng;

    if (
      latRange < SINGLE_POINT_THRESHOLD &&
      lngRange < SINGLE_POINT_THRESHOLD
    ) {
      // Single point or very tight cluster
      const centerLat = (bounds.northEast.lat + bounds.southWest.lat) / 2;
      const centerLng = (bounds.northEast.lng + bounds.southWest.lng) / 2;

      return {
        southWest: {
          lat: centerLat - SINGLE_POINT_PADDING,
          lng: centerLng - SINGLE_POINT_PADDING,
        },
        northEast: {
          lat: centerLat + SINGLE_POINT_PADDING,
          lng: centerLng + SINGLE_POINT_PADDING,
        },
      };
    }

    // Normal case - add padding
    const latPadding = latRange * NORMAL_BOUNDS_PADDING;
    const lngPadding = lngRange * NORMAL_BOUNDS_PADDING;

    return {
      southWest: {
        lat: bounds.southWest.lat - latPadding,
        lng: bounds.southWest.lng - lngPadding,
      },
      northEast: {
        lat: bounds.northEast.lat + latPadding,
        lng: bounds.northEast.lng + lngPadding,
      },
    };
  }

  /** Get padded bounds for the map component to fit */
  getPaddedBounds(): LatLngBounds | null {
    if (!this.state.conversationBounds) return null;
    return this.padBounds(this.state.conversationBounds);
  }
}

File: flutter_client/lib/bloc/map_bloc.dart
// ABOUTME: BLoC for managing map visualization state.
// ABOUTME: Handles adding/removing geographic features and map operations.

import 'package:bloc/bloc.dart';
import 'package:flutter_map/flutter_map.dart' hide MapEvent;
import 'package:latlong2/latlong.dart';

import '../models/geo_feature.dart';
import 'map_event.dart';
import 'map_state.dart';

// Auto-frame thresholds
const double _singlePointThreshold = 0.0001; // ~11 meters at equator
const double _singlePointPadding = 0.02; // ~2km view for single points
const double _normalBoundsPadding = 0.25; // 25% padding for feature collections
const double _viewAdjustmentThreshold =
    0.3; // Adjust if new features extend >30% beyond current view
const double _minRangeForCalculation = 0.000001; // Prevent division by zero

class MapBloc extends Bloc<MapEvent, MapState> {
  MapBloc() : super(MapState()) {
    on<AddGeoFeature>(_onAddGeoFeature);
    on<RemoveGeoFeature>(_onRemoveGeoFeature);
    on<ClearMap>(_onClearMap);
    on<ZoomIn>(_onZoomIn);
    on<ZoomOut>(_onZoomOut);
    on<SetZoom>(_onSetZoom);
    on<UpdateMapPosition>(_onUpdateMapPosition);
    on<DisableAutoFrame>(_onDisableAutoFrame);
    on<EnableAutoFrame>(_onEnableAutoFrame);
  }

  void _onAddGeoFeature(AddGeoFeature event, Emitter<MapState> emit) {
    final updatedFeatures = [...state.features, event.feature];
    final allPoints = _extractPointsFromFeatures(updatedFeatures);
    final newPoints = _extractPointsFromGeometry(event.feature.geometry);

    _handleAutoFrame(
      emit: emit,
      updatedFeatures: updatedFeatures,
      allPoints: allPoints,
      newPoints: newPoints,
    );
  }

  void _onRemoveGeoFeature(RemoveGeoFeature event, Emitter<MapState> emit) {
    final updatedFeatures =
        state.features.where((f) => f.id != event.featureId).toList();

    emit(state.copyWith(features: updatedFeatures));
  }

  void _handleAutoFrame({
    required Emitter<MapState> emit,
    required List<GeoFeature> updatedFeatures,
    required List<LatLng> allPoints,
    required List<LatLng> newPoints,
  }) {
    if (!state.autoFrameEnabled || newPoints.isEmpty) {
      emit(state.copyWith(features: updatedFeatures));
      return;
    }

    final allBounds = _calculateBoundsFromPoints(allPoints);
    final newBounds = _calculateBoundsFromPoints(newPoints);

    if (allBounds == null || newBounds == null) {
      emit(state.copyWith(features: updatedFeatures));
      return;
    }

    // Adjust camera if: (1) first features in conversation, or (2) new features significantly outside current view
    MapCamera? camera;
    if (state.conversationBounds == null) {
      camera = _fitBounds(allBounds);
    } else if (_shouldAdjustView(newBounds)) {
      camera = _fitBounds(allBounds);
    }

    emit(state.copyWith(
      features: updatedFeatures,
      conversationBounds: allBounds,
      center: camera?.center ?? state.center,
      zoom: camera?.zoom ?? state.zoom,
    ));
  }

  void _onClearMap(ClearMap event, Emitter<MapState> emit) {
    emit(MapState());
  }

  void _onZoomIn(ZoomIn event, Emitter<MapState> emit) {
    final newZoom = state.zoom + 1.0;
    try {
      state.mapController.move(state.center, newZoom);
    } catch (e) {
      // Map controller may not be ready yet; state will be applied when widget renders
    }
    emit(state.copyWith(
      zoom: newZoom,
      autoFrameEnabled: false,
    ));
  }

  void _onZoomOut(ZoomOut event, Emitter<MapState> emit) {
    final newZoom = state.zoom - 1.0;
    try {
      state.mapController.move(state.center, newZoom);
    } catch (e) {
      // Map controller may not be ready yet; state will be applied when widget renders
    }
    emit(state.copyWith(
      zoom: newZoom,
      autoFrameEnabled: false,
    ));
  }

  void _onSetZoom(SetZoom event, Emitter<MapState> emit) {
    try {
      state.mapController.move(state.center, event.zoom);
    } catch (e) {
      // Map controller may not be ready yet; state will be applied when widget renders
    }
    emit(state.copyWith(zoom: event.zoom));
  }

  void _onUpdateMapPosition(UpdateMapPosition event, Emitter<MapState> emit) {
    emit(state.copyWith(
      center: event.center,
      zoom: event.zoom,
    ));
  }

  List<LatLng> _extractPointsFromFeatures(List<GeoFeature> features) {
    final points = <LatLng>[];
    for (final feature in features) {
      points.addAll(_extractPointsFromGeometry(feature.geometry));
    }
    return points;
  }

  List<LatLng> _extractPointsFromGeometry(GeoJSONGeometry geometry) {
    switch (geometry) {
      case GeoJSONPoint():
        return [LatLng(geometry.latitude, geometry.longitude)];

      case GeoJSONLineString():
        return geometry.coordinates
            .map((coord) => LatLng(coord[1], coord[0]))
            .toList();

      case GeoJSONPolygon():
        // Use exterior ring (first array) for bounds calculation
        if (geometry.coordinates.isEmpty) return [];
        return geometry.coordinates[0]
            .map((coord) => LatLng(coord[1], coord[0]))
            .toList();
    }
  }

  LatLngBounds? _calculateBoundsFromPoints(List<LatLng> points) {
    if (points.isEmpty) return null;

    double minLat = points.first.latitude;
    double maxLat = points.first.latitude;
    double minLng = points.first.longitude;
    double maxLng = points.first.longitude;

    for (var point in points) {
      minLat = minLat < point.latitude ? minLat : point.latitude;
      maxLat = maxLat > point.latitude ? maxLat : point.latitude;
      minLng = minLng < point.longitude ? minLng : point.longitude;
      maxLng = maxLng > point.longitude ? maxLng : point.longitude;
    }

    return LatLngBounds(
      LatLng(minLat, minLng),
      LatLng(maxLat, maxLng),
    );
  }

  bool _shouldAdjustView(LatLngBounds newFeatureBounds) {
    if (state.conversationBounds == null) return true;

    final currentBounds = state.conversationBounds!;

    // Calculate how much of the new features are outside current bounds
    final newSW = newFeatureBounds.southWest;
    final newNE = newFeatureBounds.northEast;
    final currSW = currentBounds.southWest;
    final currNE = currentBounds.northEast;

    // If new bounds are significantly outside current bounds, adjust
    final latRange = currNE.latitude - currSW.latitude;
    final lngRange = currNE.longitude - currSW.longitude;

    // Guard against division by zero for very small or zero ranges
    // If current view is essentially a point, always adjust when new features are added
    if (latRange.abs() < _minRangeForCalculation ||
        lngRange.abs() < _minRangeForCalculation) {
      return true;
    }

    final latOutside = (newSW.latitude < currSW.latitude
            ? currSW.latitude - newSW.latitude
            : 0) +
        (newNE.latitude > currNE.latitude
            ? newNE.latitude - currNE.latitude
            : 0);
    final lngOutside = (newSW.longitude < currSW.longitude
            ? currSW.longitude - newSW.longitude
            : 0) +
        (newNE.longitude > currNE.longitude
            ? newNE.longitude - currNE.longitude
            : 0);

    // Adjust if new features extend beyond threshold of current view in any direction
    return (latOutside / latRange > _viewAdjustmentThreshold) ||
        (lngOutside / lngRange > _viewAdjustmentThreshold);
  }

  MapCamera? _fitBounds(LatLngBounds bounds) {
    try {
      final latRange = bounds.northEast.latitude - bounds.southWest.latitude;
      final lngRange = bounds.northEast.longitude - bounds.southWest.longitude;

      LatLngBounds paddedBounds;

      if (latRange < _singlePointThreshold &&
          lngRange < _singlePointThreshold) {
        // Single point or very tight cluster - create fixed bounds around it
        final center = LatLng(
          (bounds.northEast.latitude + bounds.southWest.latitude) / 2,
          (bounds.northEast.longitude + bounds.southWest.longitude) / 2,
        );
        paddedBounds = LatLngBounds(
          LatLng(center.latitude - _singlePointPadding,
              center.longitude - _singlePointPadding),
          LatLng(center.latitude + _singlePointPadding,
              center.longitude + _singlePointPadding),
        );
      } else {
        // Normal case - add padding for better visibility
        final latPadding = latRange * _normalBoundsPadding;
        final lngPadding = lngRange * _normalBoundsPadding;

        paddedBounds = LatLngBounds(
          LatLng(bounds.southWest.latitude - latPadding,
              bounds.southWest.longitude - lngPadding),
          LatLng(bounds.northEast.latitude + latPadding,
              bounds.northEast.longitude + lngPadding),
        );
      }

      // Fit camera to the bounds
      state.mapController.fitCamera(
        CameraFit.bounds(bounds: paddedBounds),
      );

      return state.mapController.camera;
    } catch (e) {
      // Map controller may not be ready yet; auto-frame will retry on next feature add
      return null;
    }
  }

  void _onDisableAutoFrame(DisableAutoFrame event, Emitter<MapState> emit) {
    emit(state.copyWith(autoFrameEnabled: false));
  }

  void _onEnableAutoFrame(EnableAutoFrame event, Emitter<MapState> emit) {
    // Recalculate bounds from all current features
    final allPoints = _extractPointsFromFeatures(state.features);
    final conversationBounds = _calculateBoundsFromPoints(allPoints);
    MapCamera? camera;
    if (conversationBounds != null) {
      camera = _fitBounds(conversationBounds);
    }

    emit(state.copyWith(
      autoFrameEnabled: true,
      conversationBounds: conversationBounds,
      center: camera?.center ?? state.center,
      zoom: camera?.zoom ?? state.zoom,
    ));
  }
}
